greymatter-io/operator
# Add optional Bugsnag error reporting using go build flags

Bugsnag at best, is irrelevant to end users, and at worst,
may introduce additional attack vectors, get flagged in
software scans, have network traffic blocked entirely, or
just generally cause pause for customers. To combat these
issues, we need to create separate dev/prod builds.
The dev build will contain the library so we can collect
error metrics during operation in test environments.
The production release will not only disable bugsnag
but will not even contain bugsnag code at all, so that
potential supply chain scans won't flag it.

The solution lies in go build flags and an error
reporter "interface" abstraction. Operator code interacts
with bugsnag library functions through the reporter
package functions. These functions essentially
shadow the bugnsag functions of the same name.
However, this allows us to conditionally compile
empty function stubs instead of the normal
definitions during non-dev builds. Go includes the
empty function stubs if it does not detect a build tag
and includes the working functions if it detects the
"dev" build tag. We pass in this tag through Docker
args.

Lastly, to support this new build pathway, the ci/cd
process expanded and changed to build both images,
push both images to quay, and run the dev image
during integration testing.

Other things to note: I included the CA certs for bugsnag's cert into the operator image as the base ubuntu image doesnt ship with any. I also corrected the pipeline agents to the 22.04. I modified the README to include instructions on how to use the dev version

To test:
First, install the operator without bugsnag (the non-dev latest version). Check the logs to ensure that `Initializing bugsnag` doesn't appear.

Then, follow the instructions in the readme to install the dev version of the operator.  Ensure that the logs contain `Initializing bugsnag`
